src/python/pants/option/options.py

```diff
@@ -410,8 +416,47 @@ def for_scope(
         :return: An OptionValueContainer representing the option values for the given scope.
         :raises pants.option.errors.ConfigValidationError: if the scope is unknown.
         """
+        builder = OptionValueContainerBuilder()
+        mutex_map = defaultdict(list)
+        parser = self.get_parser(scope)
+        scope_str = "global scope" if scope == GLOBAL_SCOPE else f"scope '{scope}'"
 
-        native_values = self.get_parser(scope).parse_args_native(self._native_parser)
+        for args, kwargs in parser.option_registrations_iter():
+            dest = kwargs["dest"]
+            val, rank = self._native_parser.get_value(
+                scope=scope, registration_args=args, registration_kwargs=kwargs
+            )
+            explicitly_set = rank > Rank.HARDCODED
+
+            # If we explicitly set a deprecated but not-yet-expired option, warn about it.
```

Comment:
The one change is that we used to check deprecations twice, once at registration time (but only to catch expired deprecations) and once at parse time. Now we do both in a single check, here, at parse time.